README.rst
=============
xs-authserver
=============


A component of a School Server like XS_, XSCE_ or DXS_.

**xs-authserver** implements a seamless web authentication service using XO
laptop registration capabilities.  It is heavily inspired by the Moodle OLPC-XS
authentication plugin.

*Continuous integration status*

.. image:: https://api.travis-ci.org/migonzalvar/xs-authserver.png
    :target: https://travis-ci.org/migonzalvar/xs-authserver

.. image:: https://coveralls.io/repos/migonzalvar/xs-authserver/badge.png
    :target: https://coveralls.io/r/migonzalvar/xs-authserver

*PyPI status*

.. image:: https://pypip.in/v/xs-authserver/badge.png
    :target: https://crate.io/packages/xs-authserver/

.. image:: https://pypip.in/d/xs-authserver/badge.png
    :target: https://crate.io/packages/xs-authserver/


.. _XS: http://wiki.laptop.org/go/School_server

.. _XSCE: http://schoolserver.org/

.. _DXS: https://sugardextrose.org/projects/dxs/



HISTORY.rst
===============
Release history
===============

0.1.7 (2014-02-13)
==================

- Change init database connection handle to be more compatible with previous
  versions of Flask. Specifically avoid app_context().

0.1.6 (2014-02-12)
==================

- Don't throw error if database already created


test_authserver.py
import json
import os
import sqlite3
import unittest
import tempfile

from werkzeug.datastructures import Headers

import authserver


FIXTURES = {
    'OLPC_XS_DB': """
CREATE TABLE "laptops" (
    serial VARCHAR(20) NOT NULL,
    nickname VARCHAR(200) NOT NULL,
    full_name VARCHAR(100) NOT NULL,
    pubkey TEXT NOT NULL,
    uuid VARCHAR(100),
    lastmodified TEXT DEFAULT '1970-11-12 12:34:56',
    class_group INTEGER,
    PRIMARY KEY (serial)
);
INSERT INTO "laptops" VALUES(
    'SHC23800059',
    'miguel',
    '',
    'AAAAB3NzaC1kc3MAAACBALLTqNsDS7t1/GRI6V9hrltCESM7c7Isw+OMieeL6tbCOb/hFlfnuWpreXwaN+IEp9kLbH/pH0U6fpqBbTxFkQ7sKTEm52fIBFIQinqayRqspP4MZp/+IVUZ4F/1N/E2jXpygpwciK+RaYt0T9f7aN1wp0RA7moVGwZ+W6wn6T6zAAAAFQCfuMPxCBN0cs5octiKcYDdIurRuQAAAIAU+K12HJ9p8ua3laen8rVSfSyYL5LMSWALflK6BEcZE7SiMMa1P3wOBFC21HBTZt28CyVhPOFt/6ptVsuzZeVP12141gcBSDPDE0zgGYW2ev7koL5GhiSUCn+Ag3ISjYJ3GHO3bmvQUXDV/vz9ADVX/k9xRdKaHpSx9Gmo5I5MkwAAAIAIT5Pg//ye+Bv23lMo24O2Axz5y2IlpZGKYy+pfOrwjInuPNhtLOAHn1Woq1CEwEqHEvHBxPHKB1P38rmZgajsVNu0OBEQjGwWfCMZa4IZh2+5FxpYFQnOuwDn44cexruLOH2vYHIb9FabTnRGn3XgNVSX86x3gVc7nN/Z1dBS8w==',
    '74EDF4DF-9E81-FF4F-F053-2174AACBFB86',
    '2013-09-19 11:14:34',
    3
);
""",
    'DATABASE': """
CREATE TABLE users (
    uuid VARCHAR(36) NOT NULL,
    nickname VARCHAR(200) NOT NULL,
    pkey_hash VARCHAR(40) NOT NULL,
    PRIMARY KEY (uuid),
    UNIQUE (pkey_hash)
);
INSERT INTO "users" VALUES(
    '09c131b2-25dd-11e3-89c5-e89a8f08bf39',
    'fulano',
    'bc040eb5294c5fe63f5cfd28d6961c7db6b9a2bc'
);
"""
}


class AuthserverTestCase(unittest.TestCase):
    def setUp(self):
        self.db_fds = {}
        for db, sql in FIXTURES.iteritems():
            fd, filename = tempfile.mkstemp()
            conn = sqlite3.connect(filename)
            conn.executescript(sql)
            conn.close()
            self.db_fds[db], authserver.app.config[db] = fd, filename
        authserver.app.config['TESTING'] = True
        self.app = authserver.app.test_client()

    def tearDown(self):
        for fd in self.db_fds.values():
            os.close(fd)
        os.unlink(authserver.app.config['DATABASE'])
        os.unlink(authserver.app.config['OLPC_XS_DB'])

    @staticmethod
    def make_cookie(pkey_hash):
        """Create a the magic XO cookie.

        :rtype : dict
        """
        cookie_value = json.dumps(dict(pkey_hash=pkey_hash))
        cookie = "xoid={cookie_value}".format(cookie_value=cookie_value)
        return cookie

    def test_unregistered(self):
        rv = self.app.get('/')
        assert "Please register your laptop" in rv.data
        assert "pkey_hash: None" in rv.data

    def test_registered_with_account(self):
        nickname, pkey_hash = 'fulano', 'bc040eb5294c5fe63f5cfd28d6961c7db6b9a2bc'
        cookie = self.make_cookie(pkey_hash)
        headers = Headers({'Cookie': cookie})
        rv = self.app.get('/', headers=headers)
        assert "Hello {}".format(nickname) in rv.data
        assert "pkey_hash: {}".format(pkey_hash) in rv.data

    def test_registered_without_account(self):
        nickname, pkey_hash = 'miguel', '195ed98f4975ebccb1e699d8636278b64e9276b3'
        cookie = self.make_cookie(pkey_hash)
        headers = Headers({'Cookie': cookie})
        rv = self.app.get('/', headers=headers)
        assert "Hello {}".format(nickname) in rv.data
        assert "pkey_hash: {}".format(pkey_hash) in rv.data

    def test_user_agent_sugar(self):
        user_agent = r'SugarLabs/0.98'
        headers = Headers({'User-Agent': user_agent})
        rv = self.app.get('/', headers=headers)
        assert "Detected Sugar platform" in rv.data


if __name__ == '__main__':
    unittest.main()


test_xs_authserver.py
import json
import os
import sqlite3
import unittest
import tempfile

from werkzeug.datastructures import Headers

import xs_authserver


FIXTURES = {
    'OLPC_XS_DB': """
CREATE TABLE "laptops" (
    serial VARCHAR(20) NOT NULL,
    nickname VARCHAR(200) NOT NULL,
    full_name VARCHAR(100) NOT NULL,
    pubkey TEXT NOT NULL,
    uuid VARCHAR(100),
    lastmodified TEXT DEFAULT '1970-11-12 12:34:56',
    class_group INTEGER,
    PRIMARY KEY (serial)
);
INSERT INTO "laptops" VALUES(
    'SHC23800059',
    'miguel',
    '',
    'AAAAB3NzaC1kc3MAAACBALLTqNsDS7t1/GRI6V9hrltCESM7c7Isw+OMieeL6tbCOb/hFlfnuWpreXwaN+IEp9kLbH/pH0U6fpqBbTxFkQ7sKTEm52fIBFIQinqayRqspP4MZp/+IVUZ4F/1N/E2jXpygpwciK+RaYt0T9f7aN1wp0RA7moVGwZ+W6wn6T6zAAAAFQCfuMPxCBN0cs5octiKcYDdIurRuQAAAIAU+K12HJ9p8ua3laen8rVSfSyYL5LMSWALflK6BEcZE7SiMMa1P3wOBFC21HBTZt28CyVhPOFt/6ptVsuzZeVP12141gcBSDPDE0zgGYW2ev7koL5GhiSUCn+Ag3ISjYJ3GHO3bmvQUXDV/vz9ADVX/k9xRdKaHpSx9Gmo5I5MkwAAAIAIT5Pg//ye+Bv23lMo24O2Axz5y2IlpZGKYy+pfOrwjInuPNhtLOAHn1Woq1CEwEqHEvHBxPHKB1P38rmZgajsVNu0OBEQjGwWfCMZa4IZh2+5FxpYFQnOuwDn44cexruLOH2vYHIb9FabTnRGn3XgNVSX86x3gVc7nN/Z1dBS8w==',
    '74EDF4DF-9E81-FF4F-F053-2174AACBFB86',
    '2013-09-19 11:14:34',
    3
);
""",
    'DATABASE': """
INSERT INTO "users" VALUES(
    '09c131b2-25dd-11e3-89c5-e89a8f08bf39',
    'fulano',
    'bc040eb5294c5fe63f5cfd28d6961c7db6b9a2bc'
);
"""
}


class XSAuthserverTestCase(unittest.TestCase):
    def setUp(self):
        for db, sql in FIXTURES.iteritems():
            filename = tempfile.NamedTemporaryFile(delete=False)
            filename.close()
            xs_authserver.app.config[db] = filename.name
            if db == 'DATABASE':
                xs_authserver.init_db()
            conn = sqlite3.connect(filename.name)
            conn.executescript(sql)
            conn.close()

        xs_authserver.app.config['TESTING'] = True
        self.app = xs_authserver.app.test_client()

    def tearDown(self):
        os.unlink(xs_authserver.app.config['DATABASE'])
        os.unlink(xs_authserver.app.config['OLPC_XS_DB'])

    @staticmethod
    def make_header(pkey_hash=None, extra_headers={}, **kwargs):
        headers = {}
        if pkey_hash:
            cookie_value = json.dumps(dict(pkey_hash=pkey_hash))
            cookie = "xoid={cookie_value}".format(cookie_value=cookie_value)
            headers['Cookie'] = cookie
        headers.update(extra_headers)
        headers.update(kwargs)
        return Headers(headers)

    def test_unregistered(self):
        rv = self.app.get('/')
        assert "Please register your laptop" in rv.data
        assert "pkey_hash: None" in rv.data

    def test_registered_with_account(self):
        nickname, pkey_hash = 'fulano', 'bc040eb5294c5fe63f5cfd28d6961c7db6b9a2bc'
        headers = self.make_header(pkey_hash)
        rv = self.app.get('/', headers=headers)
        assert "Hello {}".format(nickname) in rv.data
        assert "pkey_hash: {}".format(pkey_hash) in rv.data

    def test_registered_without_account(self):
        nickname, pkey_hash = 'miguel', '195ed98f4975ebccb1e699d8636278b64e9276b3'
        headers = self.make_header(pkey_hash)
        rv = self.app.get('/', headers=headers)
        assert "Hello {}".format(nickname) in rv.data
        assert "pkey_hash: {}".format(pkey_hash) in rv.data

    def test_user_agent_sugar(self):
        user_agent = r'SugarLabs/0.98'
        headers = self.make_header(extra_headers={'User-Agent': user_agent})
        rv = self.app.get('/', headers=headers)
        assert "Detected Sugar platform" in rv.data

    def test_bad_cookie(self):
        headers = self.make_header(extra_headers={'Cookie': 'xoid=badcookie'})
        rv = self.app.get('/', headers=headers)
        assert "Please register your laptop" in rv.data
        assert "pkey_hash: None" in rv.data

    def test_re_init_b(self):
        # Recreating database doesn't throw an error
        assert xs_authserver.init_db()


if __name__ == '__main__':
    unittest.main()


authserver/authserver.py
import hashlib
import json
import os.path
import re
import sqlite3

from flask import Flask, request, render_template, g


OLPC_XS_DB = '/home/idmgr/identity.db'
DATABASE = os.path.abspath(os.path.join(os.path.dirname(__file__), 'test.db'))


def connect_to_database(database):
    return sqlite3.connect(database)


def get_db():
    db = getattr(g, '_database', None)
    if db is None:
        db = g._database = connect_to_database(app.config['DATABASE'])
        db.row_factory = sqlite3.Row
    return db


def get_olpc_xs_db():
    db = getattr(g, '_olpc_xs_db', None)
    if db is None:
        db = g._olpc_xs_db = connect_to_database(app.config['OLPC_XS_DB'])
        db.row_factory = sqlite3.Row
    return db


def query_db(query, args=(), one=False):
    cur = get_db().execute(query, args)
    rv = cur.fetchall()
    cur.close()
    return (rv[0] if rv else None) if one else rv


class Idmgr(object):
    """Laptops registered on idmagr.

    Correspond to a register in table ``laptop`` on idmgr database. The
    schema is defined as:

    .. code:: sql

        CREATE TABLE laptops (
            serial VARCHAR(20) NOT NULL,
            nickname VARCHAR(200) NOT NULL,
            full_name VARCHAR(100) NOT NULL,
            pubkey TEXT NOT NULL,
            uuid VARCHAR(100),
            lastmodified TEXT DEFAULT '1970-11-12 12:34:56',
            class_group INTEGER,
            PRIMARY KEY (serial)
        );
    """
    fields = ("serial", "nickname", "full_name", "pubkey")

    @classmethod
    def all(cls):
        query = "SELECT {} FROM laptops".format(",".join(cls.fields))
        cur = get_olpc_xs_db().execute(query)
        rv = cur.fetchall()
        cur.close()
        return [cls(**dict(zip(r.keys(), r))) for r in rv]

    def __init__(self, **kwargs):
        self.values = {k: kwargs[k] for k in kwargs if k in self.fields}

    def __getattr__(self, name):
        if name in self.fields:
            return self.values.get(name, None)
        raise AttributeError("%r object has no attribute %r" %
                             (self.__class__, name))

    @property
    def pkey_hash(self):
        return hashlib.sha1(self.pubkey).hexdigest()

    def __repr__(self):
        d = {'pkey_hash': self.pkey_hash}
        d.update(self.values)
        return repr(d)


class User(object):
    @classmethod
    def get_by_pkey_hash(self, pkey_hash):
        query = "SELECT username FROM users WHERE pkey_hash = ?"
        args = (pkey_hash, )
        cur = get_db().execute(query, args)
        rv = cur.fetchall()
        cur.close()
        return rv[0] if rv else None

    def __init__(self, username, pkey_hash):
        self.username = username
        self.pkey_hash = pkey_hash

    def save(self):
        pass



def sync_idmgr():
    for i in Idmgr.all():
        u = User(i.nickname, i.pkey_hash)
        u.save()


app = Flask(__name__)
app.debug = True

app.config.from_object(__name__)

@app.teardown_appcontext
def close_connection(exception):
    for db_name in ('_database', '_olpc_xs_db'):
        db = getattr(g, db_name, None)
        if db is not None:
            db.close()


def identify_user_agent(user_agent):
    for sugar_ua in re.findall(r'SugarLabs/[\d\.]+', user_agent):
        platform = True
        version = sugar_ua.split("/")[1]
        break
    else:
        platform, version = False, None
    return {'sugar_platform': platform, 'sugar_version': version}


@app.route('/')
def index():
    context = dict(request=request)
    context.update(identify_user_agent(request.user_agent.string))

    context.update(pkey_hash=None, user=None)
    cookie = request.cookies.get('xoid')
    if cookie:
        xoid = json.loads(cookie)
        if xoid:
            pkey_hash = xoid.get('pkey_hash')
            if pkey_hash:
                context.update(pkey_hash=pkey_hash)
                context.update(user=User.get_by_pkey_hash(pkey_hash))

    context['registered_users'] = Idmgr.all()
    return render_template("index.html", **context)


if __name__ == '__main__':
    app.run(host='0.0.0.0')


xs_authserver/__init__.py
# -*- coding: utf-8 -*-

# Copyright © 2013 Miguel González <migonzalvar@activitycentral.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>

import hashlib
import json
import os.path
import re
import sqlite3
import uuid
import sys
import logging
from contextlib import closing

from flask import Flask, request, render_template, g

# Default config
ENV_PREFIX = 'XS_AUTHSERVER'
ENV_VARS = {
    'OLPC_XS_DB': '/home/idmgr/identity.db',
    'DATABASE': os.path.abspath(os.path.join(os.path.dirname(__file__), 'test.db')),
}


def connect_to_database(database):
    return sqlite3.connect(database)


def get_db():
    db = getattr(g, '_database', None)
    if db is None:
        db = g._database = connect_to_database(app.config['DATABASE'])
        db.row_factory = sqlite3.Row
    return db


def init_db():
    with closing(connect_to_database(app.config['DATABASE'])) as db:
        try:
            with app.open_resource('schema.sql', mode='r') as f:
                db.cursor().executescript(f.read())
            db.commit()
            rv = True
        except sqlite3.OperationalError:
            rv = False
    return rv

def get_olpc_xs_db():
    db = getattr(g, '_olpc_xs_db', None)
    if db is None:
        db = g._olpc_xs_db = connect_to_database(app.config['OLPC_XS_DB'])
        db.row_factory = sqlite3.Row
    return db


class Idmgr(object):
    """Laptops registered on idmgr.

    Correspond to a register in table ``laptop`` on idmgr database. The
    schema is defined as:

    .. code:: sql

        CREATE TABLE laptops (
            serial VARCHAR(20) NOT NULL,
            nickname VARCHAR(200) NOT NULL,
            full_name VARCHAR(100) NOT NULL,
            pubkey TEXT NOT NULL,
            uuid VARCHAR(100),
            lastmodified TEXT DEFAULT '1970-11-12 12:34:56',
            class_group INTEGER,
            PRIMARY KEY (serial)
        );
    """
    fields = ("serial", "nickname", "full_name", "pubkey")

    @classmethod
    def all(cls):
        query = "SELECT {} FROM laptops".format(",".join(cls.fields))
        cur = get_olpc_xs_db().execute(query)
        rv = cur.fetchall()
        cur.close()
        return [cls(**dict(zip(r.keys(), r))) for r in rv]

    def __init__(self, **kwargs):
        self.values = {k: kwargs[k] for k in kwargs if k in self.fields}

    def __getattr__(self, name):
        if name in self.fields:
            return self.values.get(name, None)
        raise AttributeError("%r object has no attribute %r" %
                             (self.__class__, name))

    @property
    def pkey_hash(self):
        return hashlib.sha1(self.pubkey).hexdigest()

    def __repr__(self):
        d = {'pkey_hash': self.pkey_hash}
        d.update(self.values)
        return repr(d)


class User(object):
    """User registered in xs-authserver custom database.

    .. code:: sql

        CREATE TABLE users (
            uuid VARCHAR(36) NOT NULL,
            nickname VARCHAR(200) NOT NULL,
            pkey_hash VARCHAR(40) NOT NULL,
            PRIMARY KEY (uuid),
            UNIQUE (pkey_hash)
        );

    """

    @classmethod
    def _get_users_by_pkey_hash(cls, pkey_hash):
        query = "SELECT uuid FROM users WHERE pkey_hash = ?"
        args = (pkey_hash, )
        cur = get_db().execute(query, args)
        rv = cur.fetchall()
        cur.close()
        return rv

    @classmethod
    def by_pkey_hash(cls, pkey_hash):
        rv = cls._get_users_by_pkey_hash(pkey_hash)
        if rv == []:
            sync_idmgr()
            rv = cls._get_users_by_pkey_hash(pkey_hash)
        if rv:
            return cls(uuid=rv[0]['uuid'])
        else:
            return None

    def __init__(self, **kwargs):
        if 'uuid' in kwargs:
            self.uuid = kwargs['uuid']
            self._load()
        else:
            self.uuid = None
            for field in kwargs:
                if field in ('nickname', 'pkey_hash'):
                    setattr(self, field, kwargs[field])

    def _load(self):
        query = "SELECT nickname, pkey_hash FROM users WHERE uuid = ?"
        args = (self.uuid, )
        cur = get_db().execute(query, args)
        rv = cur.fetchall()
        if rv:
            self.nickname = rv[0]['nickname']
            self.pkey_hash = rv[0]['pkey_hash']
        else:
            raise LookupError("User with UUID {uuid} not found".format(uuid=self.uuid))

    def save(self):
        if self.uuid:
            uuid_ = self.uuid
            query = "REPLACE users VALUES (?, ?, ?)"
        else:
            uuid_ = str(uuid.uuid4())
            query = "INSERT INTO users VALUES (?, ?, ?)"
        args = (uuid_, self.nickname, self.pkey_hash)
        get_db().execute(query, args)
        self.uuid = uuid_


def sync_idmgr():
    """Create new accounts from idmgr."""
    for i in Idmgr.all():
        u = User(nickname=i.nickname, pkey_hash=i.pkey_hash)
        try:
            u.save()
        except sqlite3.IntegrityError:
            # pkey_hash is not unique
            pass


app = Flask(__name__)

# Config section
for var in ENV_VARS:
    env_var = ENV_PREFIX + '_' + var
    value = os.environ.get(env_var) or ENV_VARS[var]
    app.config[var] = value

if app.debug is not True:
    stdout_handler = logging.StreamHandler(sys.stdout)
    stdout_handler.setLevel(logging.WARNING)
    formatter = logging.Formatter("%(asctime)s - %(name)s - %(levelname)s - %(message)s")
    stdout_handler.setFormatter(formatter)
    app.logger.addHandler(stdout_handler)


@app.teardown_appcontext
def close_connection(exception):
    for db_name in ('_database', '_olpc_xs_db'):
        db = getattr(g, db_name, None)
        if db is not None:
            db.close()


def identify_user_agent(user_agent):
    for sugar_ua in re.findall(r'SugarLabs/[\d\.]+', user_agent):
        platform = True
        version = sugar_ua.split("/")[1]
        break
    else:
        platform, version = False, None
    return {'sugar_platform': platform, 'sugar_version': version}


@app.errorhandler(500)
def internal_error(exception):
    app.logger.exception(exception)
    return render_template('500.html'), 500


@app.route('/')
def index():
    context = dict(request=request)
    context.update(identify_user_agent(request.user_agent.string))

    context.update(pkey_hash=None, user=None)
    cookie = request.cookies.get('xoid')
    if cookie:
        try:
            xoid = json.loads(cookie)
        except ValueError:
            app.logger.warning('xoid cookie is malformed')
            xoid = {}
        if xoid:
            pkey_hash = xoid.get('pkey_hash')
            if pkey_hash:
                context.update(pkey_hash=pkey_hash)
                context.update(user=User.by_pkey_hash(pkey_hash))

    context['registered_users'] = Idmgr.all()
    return render_template("index.html", **context)
